metis-dereference/metis-dereference-service/src/main/java/eu/europeana/metis/dereference/service/MongoDereferenceService.java

```diff
@@ -166,79 +171,158 @@ private void extractBroaderResources(EnrichmentBase resource, Set<String> destin
+      enrichmentBaseVocabularyPair = null;
+    } else {
+      enrichmentBaseVocabularyPair = convertToEnrichmentBaseVocabularyPair(
+          entityVocabularyPair.getLeft(), entityVocabularyPair.getRight());
+    }
+    return enrichmentBaseVocabularyPair;
+  }
+
+  private boolean didVocabularyChange(ProcessedEntity cachedEntity,
+      Pair<String, Vocabulary> entityVocabularyPair) {
+    final String cachedVocabularyId = cachedEntity.getVocabularyId();
+    final String computedVocabularyId = Optional.ofNullable(entityVocabularyPair.getRight())
+        .map(Vocabulary::getId).map(ObjectId::toString).orElse(null);
+
+    final boolean didVocabularyChange;
+    if (cachedVocabularyId != null && computedVocabularyId != null) {
+      didVocabularyChange = !cachedVocabularyId.equals(computedVocabularyId);
+    } else {
+      didVocabularyChange = cachedVocabularyId != null || computedVocabularyId != null;
+    }
+    return didVocabularyChange;
+  }
+
+  /**
+   * Computes the entity and vocabulary.
+   * <p>It will use the cache if it's still valid, otherwise it will retrieve(if applicable) the
+   * original entity and transform the result. </p>
+   * <p>The possible outcomes are:
+   * <ul>
+   *   <li>Both items of the pair are null. We do not have a vocabulary candidate or we have more
+   *   than one vocabulary candidate and all have not succeed either retrieving the original
+   *   entity or transforming the retrieved entity.</li>
+   *   <li>Entity xml(Left) is null, and vocabulary(Right) is non null. We have a vocabulary
+   *   and the entity xml failed either to be retried or failed transformation.</li>
+   *   <li>Entity xml(Left) is non null, and vocabulary(Right) is non null. We have a
+   *   successful retrieval and transformation.</li>
+   * </ul>
+   * </p>
+   *
+   * @param resourceId the url of the provider entity
+   * @param cachedEntity the cached entity object
+   * @return a pair with the computed values
+   * @throws URISyntaxException if the resource identifier url is invalid
+   * @throws TransformerException if an exception occurred during transformation of the original
+   * entity
+   */
+  private Pair<String, Vocabulary> computeEntityVocabularyPair(String resourceId,
+      ProcessedEntity cachedEntity) throws URISyntaxException, TransformerException {
+
+    final Pair<String, Vocabulary> transformedEntityVocabularyPair;
+
+    //Check if vocabulary actually exists
+    Vocabulary cachedVocabulary = null;
+    boolean cachedVocabularyChanged = false;
+    if (cachedEntity != null && StringUtils.isNotBlank(cachedEntity.getVocabularyId())) {
+      cachedVocabulary = vocabularyDao.get(cachedEntity.getVocabularyId());
+      cachedVocabularyChanged = cachedVocabulary == null;
+    }
+
+    // If we do not have any cached entity, we need to compute it
+    if (cachedEntity == null || cachedVocabularyChanged) {
+      transformedEntityVocabularyPair = retrieveAndTransformEntity(resourceId);
```

Comment:
I think (but please verify) that if you just add a line saying 
```
saveEntity(resourceId, cachedEntity, transformedEntityVocabularyPair)
```
after this line, you can remove the whole `didVocabularyChange()` method. You're essentially checking the exact same condition here, right? What do you think?

Reply:
Yes indeed! But I thought that on your previous comment you wanted the saving functionality outside of the method.
I can add it back in the method if you want

Reply:
First of all, I didn't say that :innocent: but secondly: you did, it is now in it's own method... I think it would be good to put it back especially since it doesn't add complexity (it's just one method call now). Does that sound OK?

Reply:
That's how I interpreted it.. :sweat_smile: 
I moved that call inside the method